File: src/sooljang/api/schemas/discovery.py
"""새 탐색은 식별 속성만 입력받는다. 자유 provider 옵션이나 개인 이력은 받지 않는다."""

import uuid
from typing import Literal

from pydantic import AwareDatetime, BaseModel, ConfigDict, Field, field_validator

from sooljang.api.schemas.interests import InterestIdentity, InterestSourceMatch


class SearchInput(BaseModel):
    model_config = ConfigDict(extra="forbid")
    connection_id: uuid.UUID
    query: str = Field(min_length=1, max_length=300)
    page: int = Field(default=1, ge=1, le=3)
    language: Literal["ko", "en", "ja", "zh-hans"] = "ko"

    @field_validator("query")
    @classmethod
    def query_nonblank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("검색명을 입력하세요")
        return value.strip()


class IdentityLookupInput(BaseModel):
    model_config = ConfigDict(extra="forbid")
    identity: InterestIdentity
    source_ids: list[uuid.UUID] = Field(min_length=1, max_length=4)
    source_matches: dict[uuid.UUID, InterestSourceMatch] = Field(default_factory=dict, max_length=4)


class InterestLookupInput(BaseModel):
    model_config = ConfigDict(extra="forbid")
    source_ids: list[uuid.UUID] = Field(min_length=1, max_length=4)


class ApplyEvidenceInput(BaseModel):
    model_config = ConfigDict(extra="forbid")
    product_id: uuid.UUID
    expected_updated_at: AwareDatetime
    evidence_token: str = Field(min_length=1, max_length=16000)
    selected_fields: list[
        Literal["name_en", "country", "region", "abv", "vintage", "age_years"]
    ] = Field(min_length=1, max_length=6)

File: web/src/api/discovery.ts
import { request } from "@/api/client";
import type { Interest, InterestIdentity, InterestSourceMatch } from "@/api/interests";
import type { SourceLookupResult, SourceOutcome } from "@/api/types";

export type ApplyField = "name_en" | "country" | "region" | "abv" | "vintage" | "age_years";
export type ApplicableFields = Partial<Record<ApplyField, string | number>>;
export interface Evidence {
  applicable_fields?: ApplicableFields;
  evidence_token?: string | null;
}
export interface DiscoveryDocument extends Evidence {
  id: string;
  url: string;
  domain: string;
  title: string;
  excerpt: string;
  fetched_at: string;
  published_at: string | null;
  evidence: "search_excerpt" | "provider_text";
  kind: "info" | "review" | "seller_note" | "customer_tasting" | "service_review" | "unknown";
  discovered_by: string[];
  rating: number | null;
  rating_scale: number | null;
  rating_count: number | null;
}
export interface SearchResponse {
  connection_id: string;
  outcome: SourceOutcome;
  documents: DiscoveryDocument[];
  warning: string | null;
  requests: number;
}
export type DiscoverySource = SourceLookupResult & Evidence;
export type SourceMatch = InterestSourceMatch;
export const discoveryApi = {
  search: (connection_id: string, query: string, signal: AbortSignal) =>
    request<SearchResponse>("/discovery/search", {
      method: "POST",
      body: { connection_id, query, page: 1, language: "ko" },
      signal,
      cache: "no-store",
    }),
  lookup: (
    identity: InterestIdentity,
    source_id: string,
    source_matches: Record<string, SourceMatch>,
    signal: AbortSignal,
  ) =>
    request<DiscoverySource[]>("/discovery/lookup", {
      method: "POST",
      body: { identity, source_ids: [source_id], source_matches },
      signal,
      cache: "no-store",
    }),
  productContext: (product_id: string, signal: AbortSignal) =>
    request<{
      identity: InterestIdentity;
      source_matches: Record<string, SourceMatch>;
      updated_at: string;
    }>(`/discovery/products/${product_id}/context`, { signal, cache: "no-store" }),
  interestLookup: (interest_id: string, source_id: string, signal: AbortSignal) =>
    request<DiscoverySource[]>(`/discovery/interests/${interest_id}/lookup`, {
      method: "POST",
      body: { source_ids: [source_id] },
      signal,
      cache: "no-store",
    }),
  productLookup: (product_id: string, source_id: string, signal: AbortSignal) =>
    request<DiscoverySource[]>(`/discovery/products/${product_id}/lookup`, {
      method: "POST",
      body: { source_ids: [source_id] },
      signal,
      cache: "no-store",
    }),
  apply: (
    product_id: string,
    expected_updated_at: string,
    evidence_token: string,
    selected_fields: ApplyField[],
  ) =>
    request<{
      product_id: string;
      updated_at: string;
      applied_fields: ApplicableFields;
      source_url: string;
    }>("/discovery/apply", {
      method: "POST",
      body: { product_id, expected_updated_at, evidence_token, selected_fields },
      cache: "no-store",
    }),
  saveInterest: (
    identity: InterestIdentity,
    source_matches: Record<string, SourceMatch>,
    request_id: string,
  ) =>
    request<Interest>("/interests", {
      method: "POST",
      body: { identity, source_matches, request_id },
    }),
};
